src/wasm.h

```diff
@@ -88,25 +88,37 @@ struct Name : public cashew::IString {
   }
 };
 
-// An index in a wasm module
-typedef uint32_t Index;
-
-// An address in linear memory. For now only wasm32
-struct Address {
-  typedef uint32_t address_t;
-  address_t addr;
-  Address() : addr(0) {}
-  Address(uint64_t a) : addr(static_cast<address_t>(a)) {
-    assert(a <= std::numeric_limits<address_t>::max());
-  }
-  Address& operator=(uint64_t a) {
-    assert(a <= std::numeric_limits<address_t>::max());
-    addr = static_cast<address_t>(a);
+// An index in a wasm module. This could be e.g. a local index or an index in
+// linear memory (i.e. an address).
+template <typename T>
+struct WasmIndex {
+  typedef T value_type;
+  T addr;
+  WasmIndex() : addr(0) {}
+  static void CheckValid(uint64_t v) {
+    // Allow -1 as an invalid sentinel value
+    assert(v <= std::numeric_limits<T>::max() ||
+	   v == static_cast<uint64_t>(-1));
+  }
+  WasmIndex(uint64_t a) : addr(static_cast<T>(a)) {
```

Comment:
Yes. The semantics are (assuming that `T` is u32) that the assigned value is promoted to u64, the u64 value is checked, and then truncated for the initialization. (well technically speaking the member initialization happens first). Of course that can all get optimized away when the assigmnent is from a u32. When the assignment is from a 64 bit type, then it actually matters.